pulsar-package-management/bookkeeper-storage/src/main/java/org/apache/pulsar/packages/management/storage/bookkeeper/DLOutputStream.java

```diff
@@ -50,27 +48,27 @@ static CompletableFuture<DLOutputStream> openWriterAsync(DistributedLogManager d
         return distributedLogManager.openAsyncLogWriter().thenApply(w -> new DLOutputStream(distributedLogManager, w));
     }
 
-    private CompletableFuture<List<LogRecord>> getRecords(InputStream inputStream) {
-        CompletableFuture<List<LogRecord>> future = new CompletableFuture<>();
-        CompletableFuture.runAsync(() -> {
-            byte[] readBuffer = new byte[8192];
-            List<LogRecord> records = new ArrayList<>();
-            try {
-                int read = 0;
-                while ((read = inputStream.read(readBuffer)) != -1) {
-                    log.info("write something into the ledgers offset: {}, length: {}", offset, read);
-                    ByteBuf writeBuf = Unpooled.copiedBuffer(readBuffer, 0, read);
-                    offset += writeBuf.readableBytes();
-                    LogRecord record = new LogRecord(offset, writeBuf);
-                    records.add(record);
-                }
-                future.complete(records);
-            } catch (IOException e) {
-                log.error("Failed to get all records from the input stream", e);
-                future.completeExceptionally(e);
+    private void writeAsyncHelper(InputStream is, CompletableFuture<DLOutputStream> result) {
+        byte[] readBuffer = new byte[8192];
```

Comment:
Since you use this method to write multiple times, each time will allocate this buffer,  we can reuse it to avoid allocating the heap memory each time.